include/pmacc/mappings/kernel/MapperConcept.hpp

```diff
@@ -0,0 +1,114 @@
+ *
+ * This file is part of PMacc.
+ *
+ * PMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#pragma once
+
+#include <pmacc/types.hpp>
+
+
+namespace pmacc
+{
+    /** Concept for mapper from block indices to supercells in the given area for alpaka kernels
+     *
+     * The mapping covers the case of supercell-level parallelism between alpaka blocks.
+     * The supercells can be processed concurrently in any order.
+     * This class is not used directly, but defines a concept for such mappers.
+     *
+     * Mapper provides a 1:1 mapping from supercells in the area to alpaka blocks.
+     * (Since it is 1:1, the mapping is invertible, but only this direction is provided.)
+     * Dimensionality of the area indices and block indices is the same.
+     * A kernel must be launched with exactly getGridDim() blocks.
+     * Each block must process a single supercell with index getSuperCellIndex(blockIndex).
+     * Implementation is optimized for the standard areas in AreaType.
+     *
+     * In-block parallelism is independent of this mapping and is done by a kernel.
+     * Naturally, this parallelism should also define block size, again independent of this mapping.
+     *
+     * This pattern is used in most kernels in particle-mesh codes.
+     * Two most common parallel patterns are:
+     *    - for particle or particle-grid operations:
+     *      alpaka block per supercell with this mapping, thread-level parallelism between particles of a frame
+     *    - for grid operations:
+     *      alpaka block per supercell with this mapping, thread-level parallelism between cells of a supercell
+     *
+     * @tparam T_areaType parameter describing area to be mapped (depends on mapper type)
+     * @tparam T_MappingDescription mapping description type, base class for MapperConcept
+     * @tparam T_dim dimensionality of area and block indices
+     * @tparam T_SupercellSize compile-time supercell size
+     */
+    template<
+        uint32_t T_areaType,
+        template<unsigned, typename>
+        typename T_MappingDescription,
+        unsigned T_dim,
+        typename T_SupercellSize>
+    class MapperConcept : public T_MappingDescription<T_dim, T_SupercellSize>
```

Comment:
If I see it correctly `MapperConcept` is not used anywhere. Is this intended?
I assumed that all mappers should inherit the concept class.

Reply:
It is the same as field update functor concept we discussed some time ago. This class is only for documentation, so that if some user/developer sees template `T_Mapper` they know what interface to expect. Also to give such interface a name

Reply:
If all mapper inherits from this concept we can take care the developer is implementing all methods and that nobody is changing the interfaces and forget this concept class.
Another way is that you wrap the user mapper into this wrapper class to guarantee that all interfaces are implemented.

Reply:
Discussed offline. We now have a few of those "concept" types that are for documenting expected template interface. We should be making so that those are somehow used in the code: as wrappers, or maybe as alpaka's `ImplementsConcept`, or somehow else